logging/logging.go

```diff
@@ -1,25 +1,176 @@
 package logging
 
 import (
+	"context"
 	"encoding/json"
 	"io"
 
+	"github.com/google/uuid"
 	"go.uber.org/zap"
 	"go.uber.org/zap/zapcore"
 )
 
-func NewLogger(service string) *zap.SugaredLogger {
+type Logger struct {
+	*zap.SugaredLogger
+	id     RequestID
+	values map[string]interface{}
+}
+
+type RequestID string
+type contextKey string
+
+const (
+	RequestIDKey = contextKey("request-id")
+	LoggerKey    = contextKey("logger")
+)
+
+func NewRequestID() RequestID {
+	return RequestID(uuid.New().String())
+}
+
+func (r RequestID) String() string {
+	return string(r)
+}
+
+func (logger Logger) WithRequestID(id RequestID) Logger {
+	if id == "" {
+		logger.Warn("Request ID is empty")
+		return logger
+	}
+	if logger.id != "" {
+		logger.Warnw("Request ID already set in logger", "current request-id", logger.id, "new request-id", id)
+		return logger
+	}
+
+	return Logger{SugaredLogger: logger.With("request-id", id),
+		id: id}
+}
+
+func (logger Logger) WithResource(resourceType, name, variant string) Logger {
+	if resourceType == "" {
+		logger.Warn("Resource type is empty")
+	}
+	if name == "" {
+		logger.Warn("Resource name is empty")
+	}
+	if variant == "" {
+		logger.Warn("Resource variant is empty")
+	}
+	resourceValues := map[string]interface{}{"resource-type": resourceType, "resource-name": name, "resource-variant": variant}
+	return Logger{
+		SugaredLogger: logger.With("resource-type", resourceType, "name", name, "variant", variant),
+		id:            logger.id,
+		values:        resourceValues,
+	}
+}
+
+func (logger Logger) WithProvider(providerType, providerName string) Logger {
```

Comment:
Since it seems common to not have provider type set, we should to two Withs.

if providerType != "" {
   logger = logger.With(...
} else {
   logger.Warn
}